File: client-android/serenada-core/src/main/java/app/serenada/core/call/OpusDtxSdp.kt
package app.serenada.core.call

private fun withDtxParameter(parameters: String): String {
    val values = parameters.split(';').map(String::trim).filter(String::isNotEmpty).toMutableList()
    val dtxIndex = values.indexOfFirst { value ->
        value.substringBefore('=').trim().equals("usedtx", ignoreCase = true)
    }
    if (dtxIndex >= 0) values[dtxIndex] = "usedtx=1" else values.add("usedtx=1")
    return values.joinToString(";")
}

/** Advertise Opus DTX as the local receive preference in every audio media section. */
internal fun enableOpusDtxInSdp(sdp: String): String {
    val separator = if (sdp.contains("\r\n")) "\r\n" else "\n"
    val hasTrailingSeparator = sdp.endsWith(separator)
    val body = if (hasTrailingSeparator) sdp.dropLast(separator.length) else sdp
    val lines = if (body.isEmpty()) mutableListOf() else body.split(Regex("\\r?\\n")).toMutableList()
    val mediaStarts = lines.indices.filter { lines[it].startsWith("m=") }

    for (mediaIndex in mediaStarts.indices.reversed()) {
        val start = mediaStarts[mediaIndex]
        if (!lines[start].startsWith("m=audio ", ignoreCase = true)) continue
        val end = mediaStarts.getOrNull(mediaIndex + 1) ?: lines.size
        val opusPayloads = (start until end).mapNotNull { lineIndex ->
            val match = OPUS_RTPMAP.matchEntire(lines[lineIndex]) ?: return@mapNotNull null
            OpusPayload(match.groupValues[1], lineIndex)
        }

        for (opus in opusPayloads.asReversed()) {
            val fmtpPattern = Regex("^a=fmtp:${Regex.escape(opus.payloadType)}\\s+(.*)$", RegexOption.IGNORE_CASE)
            val fmtpIndex = (start until end).firstOrNull { fmtpPattern.matches(lines[it]) }
            if (fmtpIndex != null) {
                val parameters = fmtpPattern.matchEntire(lines[fmtpIndex])?.groupValues?.get(1).orEmpty()
                lines[fmtpIndex] = "a=fmtp:${opus.payloadType} ${withDtxParameter(parameters)}"
            } else {
                lines.add(opus.lineIndex + 1, "a=fmtp:${opus.payloadType} usedtx=1")
            }
        }
    }

    val result = lines.joinToString(separator)
    return if (hasTrailingSeparator) result + separator else result
}

private data class OpusPayload(val payloadType: String, val lineIndex: Int)

private val OPUS_RTPMAP = Regex(
    "^a=rtpmap:(\\d+)\\s+opus/48000(?:/2)?\\s*$",
    RegexOption.IGNORE_CASE,
)

File: client-ios/SerenadaCore/Sources/Call/OpusDtxSdp.swift
import Foundation

private func withDtxParameter(_ parameters: String) -> String {
    var values = parameters.split(separator: ";").map { $0.trimmingCharacters(in: .whitespaces) }.filter { !$0.isEmpty }
    if let index = values.firstIndex(where: {
        $0.split(separator: "=", maxSplits: 1).first?.trimmingCharacters(in: .whitespaces).lowercased() == "usedtx"
    }) {
        values[index] = "usedtx=1"
    } else {
        values.append("usedtx=1")
    }
    return values.joined(separator: ";")
}

/// Advertise Opus DTX as the local receive preference in every audio media section.
func enableOpusDtxInSdp(_ sdp: String) -> String {
    let separator = sdp.contains("\r\n") ? "\r\n" : "\n"
    let hasTrailingSeparator = sdp.hasSuffix(separator)
    let body = hasTrailingSeparator ? String(sdp.dropLast(separator.count)) : sdp
    var lines = body.isEmpty ? [] : body.components(separatedBy: separator)
    let mediaStarts = lines.indices.filter { lines[$0].hasPrefix("m=") }

    for mediaIndex in mediaStarts.indices.reversed() {
        let start = mediaStarts[mediaIndex]
        guard lines[start].lowercased().hasPrefix("m=audio ") else { continue }
        let end = mediaIndex + 1 < mediaStarts.count ? mediaStarts[mediaIndex + 1] : lines.count
        let opusPayloads: [(payloadType: String, lineIndex: Int)] = (start..<end).compactMap { lineIndex in
            let lowercased = lines[lineIndex].lowercased()
            guard lowercased.hasPrefix("a=rtpmap:"), lowercased.contains(" opus/48000/2") else { return nil }
            guard let payloadType = lines[lineIndex].dropFirst("a=rtpmap:".count).split(separator: " ").first else { return nil }
            return (String(payloadType), lineIndex)
        }

        for opus in opusPayloads.reversed() {
            let prefix = "a=fmtp:\(opus.payloadType) "
            if let fmtpIndex = (start..<end).first(where: { lines[$0].lowercased().hasPrefix(prefix.lowercased()) }) {
                let parameters = String(lines[fmtpIndex].dropFirst(prefix.count))
                lines[fmtpIndex] = prefix + withDtxParameter(parameters)
            } else {
                lines.insert("a=fmtp:\(opus.payloadType) usedtx=1", at: opus.lineIndex + 1)
            }
        }
    }

    let result = lines.joined(separator: separator)
    return hasTrailingSeparator ? result + separator : result
}
